Firmware/LowLevel/include/config_defaults.h

```diff
@@ -0,0 +1,26 @@
+#ifndef _CONFIG_H
```

Comment:
`_CONFIG_DEFAULTS_H`?

Reply:
Sorry if I had asked this before, don't remember if I talked out loud or just thought about it 😆:
Do we actually need/want to have all these macros in one include separated from the structure definition?

I've just sent you a [PR](https://github.com/Apehaenger/OpenMower/pull/2) to actually use the runtime config instead of these macros in main.cpp, and that makes these macros single-use. So the only purpose that comes to my mind is to make them easy to change for a custom build. But isn't that the opposite of what we're just trying to achieve? Aren't people supposed to set the correct values in their runtime config instead of messing with the *defaults*?

Again, I have [prepared something](https://github.com/rovo89/OpenMower/compare/use_config...rovo89:OpenMower:no_defaults_include?expand=1) to show what it would look like if the defaults were inlined into the structure definition. It still compiles, and from my point of view it simplifies the code without taking any freedom away.

Reply:
Oh yes! Changed.

Reply:
> Sorry if I had asked this before, don't remember if I talked out loud or just thought about it 😆: Do we actually need/want to have all these macros in one include separated from the structure definition?
> 
Well, it was one of your review hints that you picked static values out of the struct and asked if they shouldn't go into a macro at a common place :-/
I don't care because it's pretty simple with VSCode to hover over a macro and see their value.

BTW: Regarding const vs. macro (because we talked about it): Did a quick check while integrating your [LittleFS PR](https://github.com/Apehaenger/OpenMower/pull/1). The filename const took 12 Bytes RAM as well as 32 Byte Flash **more** in comparison to the macro. That's a pretty clear const- no-go for embedded projects (as long as none of the other const benefits are needed)

> I've just sent you a [PR](https://github.com/Apehaenger/OpenMower/pull/2) to actually use the runtime config instead of these macros in main.cpp, and that makes these macros single-use. So the only purpose that comes to my mind is to make them easy to change for a custom build. But isn't that the opposite of what we're just trying to achieve? Aren't people supposed to set the correct values in their runtime config instead of messing with the _defaults_?
> 

> Again, I have [prepared something](https://github.com/rovo89/OpenMower/compare/use_config...rovo89:OpenMower:no_defaults_include?expand=1) to show what it would look like if the defaults were inlined into the structure definition. It still compiles, and from my point of view it simplifies the code without taking any freedom away.

Quite thanks for the PRs! I didn't recognized them in time, because I was working on the other (ROS) side.
Will merge once I'm done there, because I expect some conflicts now.

Reply:
> Well, it was one of your review hints that you picked static values out of the struct and asked if they shouldn't go into a macro at a common place :-/

Yes... I feel bad about this now 😔 Back then, it was for consistency reasons because you had most defaults in macros already. But for code simplicity (even without tools) and the arguments mentioned above, I would now strongly prefer to get rid of the include completely. Sorry again!

> The filename const took 12 Bytes RAM as well as 32 Byte Flash more in comparison to the macro.

Did you also try it with a `static const`?

Reply:
> > The filename const took 12 Bytes RAM as well as 32 Byte Flash more in comparison to the macro.
> 
> Did you also try it with a `static const`?

I could reproduce the memory difference - until I realized that the overhead really came from using `String`:

```
const String CONFIG_FILENAME = "/openmower.cfg";
RAM:   [=         ]   5.9% (used 15496 bytes from 262144 bytes)
Flash: [          ]   4.9% (used 103136 bytes from 2093056 bytes)

static const String CONFIG_FILENAME = "/openmower.cfg";
RAM:   [=         ]   5.9% (used 15496 bytes from 262144 bytes)
Flash: [          ]   4.9% (used 103136 bytes from 2093056 bytes)

-----

#define CONFIG_FILENAME "/openmower.cfg"
RAM:   [=         ]   5.9% (used 15484 bytes from 262144 bytes)
Flash: [          ]   4.9% (used 103104 bytes from 2093056 bytes)

const char *CONFIG_FILENAME = "/openmower.cfg";
RAM:   [=         ]   5.9% (used 15484 bytes from 262144 bytes)
Flash: [          ]   4.9% (used 103104 bytes from 2093056 bytes)

static const char *CONFIG_FILENAME = "/openmower.cfg";
RAM:   [=         ]   5.9% (used 15484 bytes from 262144 bytes)
Flash: [          ]   4.9% (used 103104 bytes from 2093056 bytes)
```

So just change it to `const char *CONFIG_FILENAME`. 😉

Reply:
Ahh good to know :+1: 

You're kidding, isn't it? :-)
It would be okay for a new project, but imagine how a single `const char *whatever` would look between all the other `#define whatever` lines.

Reply:
I wasn't kidding 😆 But yeah, `#define` is of course also fine.

The bigger difference would be getting rid of config_defaults.h, as I think that will simplify the code quite a bit. And again sorry for making you go the other way first. :/

Reply:
:laughing: 
Don't worry, config_defaults will get removed. I'll merge your PR. But I stuck currently in LL/HL testing, Bitfield defaults and the like... will get that ready before ;-)

I'll go to bed now, but probably you know?
I already added the constructor
```
struct HallConfig {
  HallConfig() : mode(HallMode::UNDEFINED), level(HallLevel::ACTIVE_LOW) {};
  HallMode mode : 3;
  HallLevel level : 1;
} __attribute__((packed));
```
Thought the constructor get called with empty default initializer
```HallConfig hall_configs[MAX_HALL_INPUTS] = {};  // Set all to UNDEFINED```
but I was wrong :-1: 
There must be a more clever way than writing them 10 times as initializer or adding a init loop.

Reply:
Strange, I would have thought the same, and so does ChatGPT: https://chatgpt.com/share/670c4845-58fc-8001-bd96-d38897031054
I know that it doesn't care too much about truth, but it came with the nice idea to add logging to the standard constructor to find out whether it was called.

Reply:
Ahh cool, she might be right. I tested only with aggregate initializer, and she said that zero initialization will be chosen then. Will try without it 👍.
Quite thanks 👍

Already googled a couple of minutes yesterday, but your ChatGPTs answer is the most matching and best describing one! I do need to test/ChatGPT some time 🤔